File: src/145_binary_tree_postorder_traversal/binary_tree_postorder_traversal.cpp
#include <climits>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
   public:
    vector<int> postorderTraversal(TreeNode* root) {
        deque<pair<TreeNode*, bool>> stack;
        vector<int> res;
        stack.push_back({root, false});

        while (stack.size() > 0) {
            pair<TreeNode*, bool> p = stack.back();
            stack.pop_back();
            if (p.second) {
                res.push_back(p.first->val);
            } else if (p.first) {
                stack.push_back({p.first, true});
                stack.push_back({p.first->right, false});
                stack.push_back({p.first->left, false});
            }
        }

        return res;
    }
};

// Helper to build tree from level order array (use INT_MIN for null)
TreeNode* buildTree(const vector<int>& arr) {
    if (arr.empty()) return nullptr;
    vector<TreeNode*> nodes;
    for (int v : arr) nodes.push_back(v == INT_MIN ? nullptr : new TreeNode(v));
    int pos = 1;
    for (int i = 0; i < nodes.size() && pos < nodes.size(); ++i) {
        if (nodes[i]) {
            if (pos < nodes.size()) nodes[i]->left = nodes[pos++];
            if (pos < nodes.size()) nodes[i]->right = nodes[pos++];
        }
    }
    return nodes[0];
}

int main() {
    Solution sol;
    struct Test {
        vector<int> tree;
        vector<int> expected;
    } tests[] = {
        {{1, INT_MIN, 2, 3}, {3, 2, 1}}, {{}, {}}, {{1}, {1}}, {{1, 2, 3}, {2, 3, 1}},
        {{1, INT_MIN, 2}, {2, 1}},
    };
    for (int i = 0; i < 5; ++i) {
        TreeNode* root = buildTree(tests[i].tree);
        vector<int> result = sol.postorderTraversal(root);
        cout << "Test " << (i + 1) << ": [";
        for (int v : result) cout << v << ",";
        cout << "] | [";
        for (int v : tests[i].expected) cout << v << ",";
        cout << "]" << endl;
    }
    return 0;
}

File: src/145_binary_tree_postorder_traversal/BinaryTreePostorderTraversal.java
// Definition for a binary tree node.

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Pair {
    TreeNode node;
    boolean visited;

    Pair(TreeNode node, boolean visited) {
        this.node = node;
        this.visited = visited;
    }
}

// Solution class for logic
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        Deque<Pair> stack = new ArrayDeque<>();
        List<Integer> res = new ArrayList<>();

        stack.add(new Pair(root, false));

        while (!stack.isEmpty()) {
            Pair pair = stack.pollLast();
            if (pair.visited) {
                res.add(pair.node.val);
            } else if (pair.node != null) {
                stack.add(new Pair(pair.node, true));
                stack.add(new Pair(pair.node.right, false));
                stack.add(new Pair(pair.node.left, false));
            }
        }

        return res;
    }
}

// Main/test class
public class BinaryTreePostorderTraversal {
    // Helper to build tree from array (level order)
    private static TreeNode buildTree(Integer[] arr) {
        if (arr.length == 0)
            return null;
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        TreeNode root = new TreeNode(arr[0]);
        queue.add(root);
        int i = 1;
        while (!queue.isEmpty() && i < arr.length) {
            TreeNode node = queue.poll();
            if (i < arr.length && arr[i] != null) {
                node.left = new TreeNode(arr[i]);
                queue.add(node.left);
            }
            i++;
            if (i < arr.length && arr[i] != null) {
                node.right = new TreeNode(arr[i]);
                queue.add(node.right);
            }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        Object[][] tests = {
                { new Integer[] { 1, null, 2, 3 }, java.util.Arrays.asList(3, 2, 1) },
                { new Integer[] {}, java.util.Arrays.asList() },
                { new Integer[] { 1 }, java.util.Arrays.asList(1) },
                { new Integer[] { 1, 2, 3 }, java.util.Arrays.asList(2, 3, 1) },
                { new Integer[] { 1, null, 2 }, java.util.Arrays.asList(2, 1) },
        };
        for (int i = 0; i < tests.length; ++i) {
            TreeNode root = buildTree((Integer[]) tests[i][0]);
            java.util.List<Integer> expected = (java.util.List<Integer>) tests[i][1];
            java.util.List<Integer> result = sol.postorderTraversal(root);
            System.out.println("Test " + (i + 1) + ": " + result + " | " + expected);
        }
    }
}
